### PR TITLE
SYNF-6 - Add in ACL support to minimal long template

### PR DESCRIPTION
**Pull request by thomasyu888**
This is an example template that supports ACL which creates a challenge staging project.  This replaces this PR: https://github.com/Sage-Bionetworks/synapse-formation-templates/pull/5/

This currently only supports teams and users that have already been created in Synapse.

### CODE REVIEW

File: templates/YAML/challenge_staging_site.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+- admin_acl: &admin
+  - READ
+  - DOWNLOAD
+  - UPDATE
+  - CREATE
+  - DELETE
+  - CHANGE_SETTINGS
+  - MODERATE
+  - CHANGE_PERMISSIONS
```
Comment (mfazza) [inline templates/YAML/challenge_staging_site.yaml]: Love the abstraction into the bundled acls; I think this is a good idea.  Let's make sure we're cognizant that we're leaving out a lot of [these](https://docs.synapse.org/rest/org/sagebionetworks/repo/model/ACCESS_TYPE.html)

File: templates/YAML/challenge_staging_site.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+- &admin_team
+  type: Team
+  id: 3379097
+  name: Sage Bio DREAM Curators
+
+- &service_account
+  type: User
+  id: 3330862
```
Comment (mfazza) [inline templates/YAML/challenge_staging_site.yaml]: If we follow the API, shouldn't this be called "principalId"?

File: templates/YAML/challenge_staging_site.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  name: test_thomas.yu
+
+- name: Challenge Staging Example
+  type: Project
+  acl:
+    # principal can be both a synapse user or team
+    - principal: *admin_team
+      access_type: *admin
+  children:
+    - Infrastructure
```
Comment (mfazza) [inline templates/YAML/challenge_staging_site.yaml]: "Infrastructure" is a folder, right?

File: templates/YAML/challenge_staging_site.yaml (hunk 1 above)
Comment (thomasyu888) [reply templates/YAML/challenge_staging_site.yaml]: So... The access types above are pertinent to entities, the ones I have left out are the ones requires for evaluation queues which is not currently supported.

File: templates/YAML/challenge_staging_site.yaml (hunk 3 above)
Comment (thomasyu888) [reply templates/YAML/challenge_staging_site.yaml]: Correct.

File: templates/YAML/challenge_staging_site.yaml (hunk 1 above)
Comment (mfazza) [reply templates/YAML/challenge_staging_site.yaml]: No worries.  I just wanted to make sure we were mindful of those.  Something else: these are defined in this template, so they're contained to this context.

Comment (thomasyu888) [reply templates/YAML/challenge_staging_site.yaml]: I actually abstracted out the share setting from the web client and created it in `challengeutils` [here](https://github.com/Sage-Bionetworks/challengeutils/blob/main/challengeutils/permissions.py). (It was inconvenient to know the access type bundle for each of the share settings).   That is one solution we can opt for which is to let people know that the share settings they can pick are 'view', 'edit', 'edit_and_delete', 'admin' and the abstraction is done in the client.  Thoughts?

File: templates/YAML/challenge_staging_site.yaml (hunk 2 above)
Comment (thomasyu888) [reply templates/YAML/challenge_staging_site.yaml]: @BrunoGrandePhD  had suggested it [here](https://github.com/Sage-Bionetworks/synapse-formation-templates/pull/5#discussion_r614216374).  Im open for either.

File: templates/YAML/challenge_staging_site.yaml (hunk 1 above)
Comment (BrunoGrandePhD) [reply templates/YAML/challenge_staging_site.yaml]: For myself, I often refer to the `challengeutils` code to know the breakdown for each permission level (_i.e._ `view`, `edit`, `edit_and_delete`, and `admin`). 

However, for the average user, I don't think it's important enough to justify including these anchors in every template. Now, we could have these permission level anchors in a shared YAML document that's prepended to every input YAML template, but I don't like the idea of forcing users to have to use aliases if they want to use ACLs. I think of anchors/aliases as a power user feature and should be optional. Just my 0.02.

File: templates/YAML/challenge_staging_site.yaml (hunk 2 above)
Comment (BrunoGrandePhD) [reply templates/YAML/challenge_staging_site.yaml]: I can go either way. I like the consistency of using `name` and `id` for entities and principals, but consistency with the API might be more important. I suppose that we could easily have a mapping between various types (_e.g._ `Team`, `Folder`) and the name of the ID property (_e.g._ `principalId`, `entityId`, respectively). Since consistency with the API was reiterated during the code review last Friday, I'm willing to concede on my earlier comment about using `id` everywhere.

**Review by BrunoGrandePhD — COMMENTED** [review]
I left my review as replies on existing comment threads.

File: templates/YAML/challenge_staging_site_long.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+- name: Sage Bio DREAM Curators
+  type: Team
+  id: 3379097
+- name: test_thomas.yu
+  type: User
+  id: 3330862
```
Comment (thomasyu888) [inline templates/YAML/challenge_staging_site_long.yaml]: This isn't necessary probably until you actually have to create the teams.

File: templates/long.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+# Edge cases should be described in the resource name.
+
+- name: Test Project - With Folders
+  type: Project
+  acl:
+  - principal:
+      type: Team
+      id: 3379097
+      name: Sage Bio DREAM Curators
+    principal_id: 3379097
```
Comment (thomasyu888) [inline templates/long.yaml]: I think for pre-existing teams and users, we could support `principal_id` field.  That being said, I do like the "principal" object.   One thing that we haven't talked about is the ability to support a list of  users/teams per access type.

**Comment by BrunoGrandePhD** [comment]
I propose that we close this PR and open a new one that implements ACLs in the minimal template that will get merged in #31.

**Comment by thomasyu888** [comment]
@BrunoGrandePhD I can merge in the develop branch once the other PR is created and fix merge conflicts.

File: templates/long.yaml (hunk 5 above)
Comment (BrunoGrandePhD) [reply templates/long.yaml]: @thomasyu888: Thanks for merging `main` into this PR branch! 

In the long template, I think there should be one principal per element in the array under `acl:`, which more closely matches how Synapse [models](https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/AccessControlList.html) ACLs. That said, I think we can include a feature in the short template whereby multiple principals can share the same permission level, and the expansion code would translate that into the full array of `acls`.

File: templates/long.yaml [hunk 6]
```diff
@@ -5,9 +5,32 @@
 
 - name: Test Project - With Folders
   type: Project
+  acl:
+  - principal:
```
Comment (BrunoGrandePhD) [inline templates/long.yaml]: If our design goal for the long template is to represent the [Synapse API](https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/ResourceAccess.html), then we might want to only support `principal_id`. We can delegate support for team names to the short template, whereby the principal ID would be added in during expansion as long as the team already exists. 

That said, we could still include team information in the long template as top-level items (similar to how we envision new teams will look like). This would be optional for template authors, but they might want to include this for their own reference. I imagine that this would ultimately get ignored by the deployment code since all it needs is the principal ID. 

I've included a preview of what I mean below: 

```
- name: Test Project - With Folders
  type: Project
  acl:
  - principal_id: 3379097
    access_type:
    - READ
    - DOWNLOAD
    - UPDATE
    - CREATE
    - DELETE
    - CHANGE_SETTINGS
    - MODERATE
    - CHANGE_PERMISSIONS
  children:
  - name: Test Folder - With Subfolder and One of Many
    type: Folder
    acl:
    - principal_id: 3330862
      access_type:
      - READ
      - DOWNLOAD
    children:
    - name: Test Subfolder - Empty
      type: Folder
  - name: Test Folder - Empty and One of Many
    type: Folder

- name: Test Project - Empty
  type: Project

- name: synapseformation-test-user
  type: User
  id: 3426116

- name: SynapseFormation Test Team
  type: Team
  id: 3426115
```

File: templates/long.yaml [hunk 7]
```diff
@@ -5,9 +5,32 @@
...
+    principal_id: 3379097
+    access_type:
+    - READ
+    - DOWNLOAD
+    - UPDATE
+    - CREATE
+    - DELETE
+    - CHANGE_SETTINGS
+    - MODERATE
+    - CHANGE_PERMISSIONS
```
Comment (BrunoGrandePhD) [inline templates/long.yaml:18]: I like using these atomic access types in the long template because that's how Synapse [models](https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/ResourceAccess.html) ACLs. We can support the web client shortcuts (_e.g._ `edit`, `admin`, etc.) in the short template. 

That said, the template author can always use YAML anchors/aliases in the long template. I agree that we shouldn't use them in this minimal template just to keep things simple.

**Review by BrunoGrandePhD — COMMENTED** [review]
I have a few thoughts for @Sage-Bionetworks/synapseformation.

File: templates/long.yaml (hunk 6 above)
Comment (thomasyu888) [reply templates/long.yaml]: Agree with this.

**Review by BrunoGrandePhD — APPROVED** [review]
🚀

File: templates/long.yaml (hunk 7 above)
Comment (mfazza) [reply templates/long.yaml:18]: I like the idea of supporting the web client shortcuts in the short version despite the added work.  For the long template, we should keep permissions this way, following the API.  If I was to extract a project structure from a Synapse Project, this is the representation I expect to see.

**Review by mfazza — APPROVED** [review]
💯
